### PR TITLE
32bit filters

### PR DESCRIPTION
**Pull request by stiiifff**

### CODE REVIEW

**Comment by stiiifff** [comment]
Implements #8

File: src/xor32.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        let keys: Vec<u64> = (0..SAMPLE_SIZE).map(|_| rng.gen()).collect();
+
+        let filter = Xor32::from(&keys);
+
+        let false_positives: usize = (0..SAMPLE_SIZE)
+            .map(|_| rng.gen())
+            .filter(|n| filter.contains(n))
+            .count();
+        let fp_rate: f64 = (false_positives * 100) as f64 / SAMPLE_SIZE as f64;
+        assert!(fp_rate < 0.0000000000000001, "False positive rate is {}", fp_rate);
```
Comment (stiiifff) [inline src/xor32.rs:141]: This actually passed, so 1B key sample is probably still too small.

File: src/xor32.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        let keys: Vec<u64> = (0..SAMPLE_SIZE).map(|_| rng.gen()).collect();
+
+        let filter = Xor32::from(&keys);
+        let bpe = (filter.len() as f64) * 32.0 / (SAMPLE_SIZE as f64);
+
+        assert!(bpe < 40., "Bits per entry is {}", bpe);
+    }
+
+    #[test]
+    #[ignore]
```
Comment (ayazhafiz) [inline src/xor32.rs:128]: is this ignored because of the large memory cost?

File: src/xor32.rs (hunk 1 above)
Comment (ayazhafiz) [reply src/xor32.rs:141]: Can you add a comment on this tests noting that?

File: src/fuse32.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+use alloc::{boxed::Box, vec::Vec};
+use core::convert::TryFrom;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// Xor filter using 32-bit fingerprints in a [fuse graph]. Requires less space than an [`Xor32`].
+///
+/// A `Fuse32` filter uses <36.404 bits per entry of the set is it constructed from, and has a false
+/// positive rate of <0.001%. As with other probabilistic filters, a higher number of entries decreases
```
Comment (ayazhafiz) [inline src/fuse32.rs:13]: honestly maybe we can just say
```suggestion
/// A `Fuse32` filter uses <36.404 bits per entry of the set is it constructed from, and has a false
/// positive rate of effectively zero (1/2^32 =~ 1/4 billion). As with other probabilistic filters, a higher number of entries decreases
```

File: src/fuse32.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug)]
+pub struct Fuse32 {
+    seed: u64,
+    segment_length: usize,
+    fingerprints: Box<[u32]>,
+}
+
+impl Filter<u64> for Fuse32 {
+    /// Returns `true` if the filter contains the specified key. Has a false positive rate of <0.02%.
```
Comment (ayazhafiz) [inline src/fuse32.rs:67]: Comment should be updated for false positive rate, or removed

File: src/fuse32.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+///     .filter(|n| filter.contains(n))
+///     .count();
+/// let fp_rate: f64 = (false_positives * 100) as f64 / SAMPLE_SIZE as f64;
+/// assert!(fp_rate < 0.001, "False positive rate is {}", fp_rate);
+/// ```
+///
+/// Serializing and deserializing `Fuse32` filters can be enabled with the [`serde`] feature.
+///
+/// [fuse graph]: https://arxiv.org/abs/1907.04749
+/// [`Xor16`]: crate::Xor16
```
Comment (ayazhafiz) [inline src/fuse32.rs:56]:
```suggestion
/// [`Xor32`]: crate::Xor32
```

File: src/xor32.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+/// [`serde`]: http://serde.rs
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Xor32 {
+    seed: u64,
+    block_length: usize,
+    fingerprints: Box<[u32]>,
+}
+
+impl Filter<u64> for Xor32 {
+    /// Returns `true` if the filter contains the specified key. Has a false positive rate of <0.02%.
```
Comment (ayazhafiz) [inline src/xor32.rs:59]: Comment should be updated for false positive rate, or removed

File: src/xor32.rs (hunk 2 above)
Comment (stiiifff) [reply src/xor32.rs:128]: Indeed, but also the long time it takes to run as the crypto-rng is single-threaded.

**Pull request merged**